pkg/sql/schemachanger/scbuild/build.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scbuild
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scbuild/internal/scbuildstmts"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scerrors"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/errors"
+)
+
+// Build constructs a new state from an initial state and a statement.
+//
+// The function takes an AST for a DDL statement and constructs targets
+// which represent schema changes to be performed.
+func Build(
+	ctx context.Context, dependencies Dependencies, initial scpb.State, n tree.Statement,
+) (scpb.State, error) {
+	bs := newBuilderState(dependencies, initial, n)
+	{
+		bCtx := buildCtx{
+			Dependencies: dependencies,
+			BuilderState: &bs,
+		}
+		if err := ProcessStatement(ctx, bCtx, n); err != nil {
+			return scpb.State{}, err
+		}
+	}
```

Comment:
now I'm being supremely nit-picky, but aesthetically, how do you feel about:

```go
bs := newBuilderState(dependencies, initial, n)
if err := scbuildstmt.Process(ctx, makeBuildCtx(deps, bs), stmt); err != nil {
    return scpb.State{}, err
}
return *bs.output, nil
```

Reply:
Another thought that occurs to me is that we can make `BuildCtx` embed `context.Context` and then eliminate an argument all over the place. It is a context after all. Then this would look like:

```go
if err := scbuildstmt.Process(makeBuildCtx(ctx, deps, bs), stmt); err != nil {
    return scpb.State{}, err
}
```
where
```go
func makeBuildCtx(ctx context.Context, deps Dependencies, builderState BuilderState) buildCtx {
    return buildCtx{
        Context: ctx,
        Dependencies: deps,
        BuilderState: builderState,
    }
}
```

This way all the `scbuildstmt` functions can look like:

```
func AlterTable(ctx BuildCtx, n *tree.AlterTable) {
```

I might even go one further and rename `BuildCtx` to `Context`.

Reply:
Okay, one more thought. I feel like the most compelling thing about keeping `ProcessStatement` in this package was that it clarified the error handling boundaries, namely, it clarified that all error propagation during statement building is panicked. I think we can and should maintain that. Consider:

```go
package scbuildstmt

// Process will process the provided statement. Errors are panicked.
func Process(ctx BuildCtx, n tree.Statement) {
	switch n := n.(type) {
	case *tree.AlterTable:
		scbuildstmts.AlterTable(ctx, bCtx, n)
	case *tree.CreateIndex:
		scbuildstmts.CreateIndex(ctx, bCtx, n)
	case *tree.DropDatabase:
		scbuildstmts.DropDatabase(ctx, bCtx, n)
	case *tree.DropSchema:
		scbuildstmts.DropSchema(ctx, bCtx, n)
	case *tree.DropSequence:
		scbuildstmts.DropSequence(ctx, bCtx, n)
	case *tree.DropTable:
		scbuildstmts.DropTable(ctx, bCtx, n)
	case *tree.DropType:
		scbuildstmts.DropType(ctx, bCtx, n)
	case *tree.DropView:
		scbuildstmts.DropView(ctx, bCtx, n)
	default:
		panic(scerrors.NotImplementedError(n))
	}
}
```

Then
```go
package scbuild

// Build ...
func Build(
	ctx context.Context, deps Dependencies, initial scpb.State, stmt tree.Statement,
) (_ scpb.State, err error) {
    	defer func() {
    	        switch r := recover().(type) {
    	        case nil: // no error
    	        case error:
    	            err = r
    	        default:
    	            err =  errors.AssertionFailedf(
    	                "unexpected error (%T) encountered while building schema change for %s: %s",
    	                 r, stmt, r,
    	             )
    	        }
	}()
	bs := newBuilderState(deps, initial, stmt)
	scbuildstmt.Process(makeBuildCtx(ctx, deps, bs), stmt)
	return *bs.output, nil
}
```

Reply:
Embedding Context! I like it